### PR TITLE
Create App. - Feature Bundle - Nov 2020

### PR DESCRIPTION
**Pull request by AleziaKurdis**
Contains:

1- **Entity List - Highlighting of the Last Selected entity**
The last selected entity is now displayed in darker blue in the list (if present per radius search)
This makes clearer which entity could become the parent.

2- Columns **visibility** and **order** are now **persistent**.

3- For clarity, "**Select Family**" and "**Select Top Family**" have been **renamed** respectively:
"**Select Parent And All Its Children**" and "**Select Top Parent And All Its Children**"

4- **New in-world bounding box color for a selected entity that is Parent & Child at the same time.**
When 1 entity is selected (in-world):
if the entity is a **Top Parent**, the selection color of the bounding box is **Orange**.
if the entity is **Parent and Child**, the selection color of the bounding box is **Majenta**. <=NEW!
if the entity is only a **Child**, then the selection color of the bounding box is **Cyan**.
If not involved in any parent line, or if the selection is multiple, then the selection color of the bounding box is **Light Grey**.

5- **Bug fix**: The "Selection" and "Actions" menu now close when we close the Entity list window or the Create App.

6- **Bug fix**: “Move Selected Entities to Avatar” is now recorded in the undo history. So undo/redo can be done.

7- **Bug fix**: EntityList, CTRL-Click was adding the last selected entity as the first selection the selection stack. The first selected item was systematically the last item of the selection. This was causing unpredictable results with "Parent Entities To The Last Selected". (This bug becomes evident by highlighting the "Last Selected" in the entity list.

### CODE REVIEW

**Comment by AleziaKurdis** [comment]
Please, **Do Not Merge** now.
I still have some planned stuff for this bundle.
_(But note that what is in this PR if fully functional as it is)_

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-dbbae4d_38e15b62.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-dbbae4d.tar_38e15b62.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-dbbae4d_38e15b62.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-dbbae4d.tar_38e15b62.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-dbbae4d_38e15b62.exe

**Comment by AleziaKurdis** [comment]
8- Add a new User Preference for the Default value of the Entity List Radius. This is available in the Edit menu. If changed, this value will be used as the default Radius for the next time the script will be loaded.

9- Changed the label for "URL of a .json to import" in "Edit > Import Entities from URL". (it was previously "URL of SVO to Import", SVO being replaced by json long time ago.)

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8e08e7f_7bec7cba.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8e08e7f_7bec7cba.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8e08e7f.tar_7bec7cba.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8e08e7f.tar_7bec7cba.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8e08e7f_7bec7cba.exe

**Comment by Aitolda** [comment]
For some reason I get no loading dialogue when using this build, although worlds still seem to be loading fine.  Not sure what to make of this.

**Comment by AleziaKurdis** [comment]
10- The action “**Import Entities (.json) from a URL**” has been added to the “**Create**” **Tab**. (It was available only in the Edit menu.)
image

11- The Create App. **Menu Items** under the “**Edit**” menu have been cleaned up.
All the actions, being transferred to the Create App. UI, has been removed. (Some was not even decently usable from there, mainly in HMD.)
A new-sub menu “**Create Application – Preferences**” has been implemented to regroup the remaining menu items that are all **User Preferences**. 
image

Eventually in the future, these might be moved to the Create App. UI. For now, it will be OK, it's already cleaner.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-0508efc_e0cac2b9.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-0508efc.tar_e0cac2b9.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-0508efc.tar_e0cac2b9.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-0508efc_e0cac2b9.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-0508efc_e0cac2b9.exe

**Comment by Aitolda** [comment]
[if the entity is Parent and Child, the selection color of the bounding box is Majenta](https://i.gyazo.com/32e2da10e556f01dacf907af14a5a7e1.png)

This doesn't seem to be working. I still get grey.

**Comment by AleziaKurdis** [comment]
> [if the entity is Parent and Child, the selection color of the bounding box is Majenta](https://i.gyazo.com/32e2da10e556f01dacf907af14a5a7e1.png)
> 
> This doesn't seem to be working. I still get grey.

In you screenshot, you have a multiple selection. 
The box will be light gray for multiple selection, since it can be a collection of different hierarchies.

The color will be applicable only on a single selection.
It's mainly there to give a minimal feedback about the patent state of what is selected.
so you can figure if it's a Top Parent (orange), just a parent inside a hierarchy(majenta), a child (cyan), or not involved in any hierarchy (light gray)

**Comment by Aitolda** [comment]
> > [if the entity is Parent and Child, the selection color of the bounding box is Majenta](https://i.gyazo.com/32e2da10e556f01dacf907af14a5a7e1.png)
> > This doesn't seem to be working. I still get grey.
> 
> In you screenshot, you have a multiple selection.
> The box will be light gray for multiple selection, since it can be a collection of different hierarchies.
> 
> The color will be applicable only on a single selection.
> It's mainly there to give a minimal feedback about the patent state of what is selected.
> so you can figure if it's a Top Parent (orange), just a parent inside a hierarchy(majenta), a child (cyan), or not involved in any hierarchy (light gray)

Ok.  I misunderstood. I took it to mean that if you had both a parent and a child selected it would be magenta.  Now I realize you meant a single entity which is both a parent and a child, in the middle of the hierarchy.

**Comment by AleziaKurdis** [comment]
That will be that for this feature bundle.
**It can be merged, once reviewed, tested, accepted.**

**KNOWN ISSUES:**
**#861 Menu.addMenuItem can't position item over existing ones in HMD Tablet Menu**
_This makes "Undo" and "Redo" to appear at the end of the "Edit" menu in HDM Tablet Menu only. The issue was present before this current PR. So the current script is expected to be right, but there is a bug somewhere in the code side._
Severity: Low
Priority: Low

**Comment by Aitolda** [comment]
It's working fantastic for me.  Using it currently for a build!

File: scripts/system/create/entityList/html/entityList.html
Comment (ctrlaltdavid) [inline scripts/system/create/entityList/html/entityList.html:228]: Extraneous spaces.

File: scripts/system/create/entityList/html/js/entityList.js
Comment (ctrlaltdavid) [inline scripts/system/create/entityList/html/js/entityList.js:1717]:
```suggestion
                                        swapColumns(i - 1, i);
```
And similarly in the next two lines.

File: scripts/system/create/qml/EditToolsTabView.qml
Comment (ctrlaltdavid) [inline scripts/system/create/qml/EditToolsTabView.qml:214]: Extraneous spaces.

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8f9a1d2_da2737de.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8f9a1d2.tar_da2737de.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8f9a1d2.tar_da2737de.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8f9a1d2_da2737de.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR858/Vircadia-Alpha-PR858-8f9a1d2_da2737de.exe

**Comment by vegaslon** [comment]
QA Missed the removal or export entity from this PR.

**Comment by Aitolda** [comment]
> QA Missed the removal of export entity from the menu and no addition of it to the create app.

Export can be found at the bottom of the entity list.

**Pull request merged**
